### PR TITLE
[IPv6] Fix reconcileFrontendIPConfigs()

### LINKED ISSUE #4038: Fix or reduce frequency or switch off perma-failing jobs
https://testgrid.k8s.io/provider-azure-cloud-provider-azure#cloud-provider-azure-master-ipv6-vmss-capz

Data as of June 3rd, 9:00 PM Eastern. Latest data can be found here:
http://storage.googleapis.com/k8s-metrics/failures-latest.json

Additional Context:
https://github.com/kubernetes/test-infra/issues/18600

### PR DESCRIPTION
**Pull request by lzhecheng**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
[IPv6] Fix reconcileFrontendIPConfigs(). Current logic handles lb.FrontendIPConfigurations according to Service's IP family, which is incorrect. For an IPv6 cluster, there're still some IPv4 FIPs due to Azure limitation, which will be removed. For an IPv4 cluster, all resources are of IPv4, which is not affected.

#### Which issue(s) this PR fixes:

Fixes #4038

#### Does this PR introduce a user-facing change?

```release-note
[IPv6] Fix reconcileFrontendIPConfigs(). Current logic handles lb.FrontendIPConfigurations according to Service's IP family, which is incorrect. For an IPv6 cluster, there're still some IPv4 FIPs due to Azure limitation, which will be removed. For an IPv4 cluster, all resources are of IPv4, which is not affected.

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/azure_utils.go [hunk 1]
```diff
@@ -447,27 +447,23 @@ func getResourceByIPFamily(resource string, isDualStack, isIPv6 bool) string {
 }
 
 // isFIPIPv6 checks if the frontend IP configuration is of IPv6.
-func (az *Cloud) isFIPIPv6(service *v1.Service, fip *network.FrontendIPConfiguration, isInternal bool) (bool, error) {
-	if isInternal {
```
Comment (MartinForReal) [inline pkg/provider/azure_utils.go:451]: Could you please add some comments here? because I would never know that the privateipaddress is empty when the lb is not internal lb until I checked the fip configs of existing cluster.

File: pkg/provider/azure_utils.go [hunk 2]
```diff
@@ -447,27 +447,23 @@ func getResourceByIPFamily(resource string, isDualStack, isIPv6 bool) string {
...
 	}
+
+	// Then Check public
 	var fipPIPID string
 	if fip.FrontendIPConfigurationPropertiesFormat != nil && fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress != nil {
 		fipPIPID = pointer.StringDeref(fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress.ID, "")
 	}
-	pipResourceGroup := az.getPublicIPAddressResourceGroup(service)
-	pips, err := az.listPIP(pipResourceGroup, azcache.CacheReadTypeDefault)
+	pips, err := az.listPIP(rg, azcache.CacheReadTypeDefault)
```
Comment (MartinForReal) [inline pkg/provider/azure_utils.go:466]: can we invoke get function instead?

File: pkg/provider/azure_loadbalancer.go [hunk 3]
```diff
@@ -1820,31 +1821,53 @@ func (az *Cloud) reconcileFrontendIPConfigs(clusterName string, service *v1.Serv
...
 	}
+	// Even if one IP family is not enabled for the Service, we still need to put
+	// FIPs of that IP family into lbFrontEndIPConfigNewConfigs.
 	v4Enabled, v6Enabled := getIPFamiliesEnabled(service)
 	if v4Enabled {
 		if err := handleFrontendIPConfig(false); err != nil {
 			return ownedFIPConfigs, toDeleteConfigs, changedTotal, err
 		}
+	} else {
+		lbFrontEndIPConfigNewConfigs = append(lbFrontEndIPConfigNewConfigs, newConfigs[false]...)
```
Comment (jwtty) [inline pkg/provider/azure_loadbalancer.go:1861]: What does this "else" mean?

File: pkg/provider/azure_loadbalancer.go [hunk 4]
```diff
@@ -1820,31 +1821,53 @@ func (az *Cloud) reconcileFrontendIPConfigs(clusterName string, service *v1.Serv
...
 	}
 	if v6Enabled {
 		if err := handleFrontendIPConfig(true); err != nil {
 			return ownedFIPConfigs, toDeleteConfigs, changedTotal, err
 		}
+	} else {
+		lbFrontEndIPConfigNewConfigs = append(lbFrontEndIPConfigNewConfigs, newConfigs[true]...)
```
Comment (jwtty) [inline pkg/provider/azure_loadbalancer.go:1868]: Can we make a type definition to wrap these bools? It's hard to follow what true/false mean for readers.

File: pkg/provider/azure_loadbalancer.go [hunk 5]
```diff
@@ -1820,31 +1821,53 @@ func (az *Cloud) reconcileFrontendIPConfigs(clusterName string, service *v1.Serv
...
+			if err != nil {
+				return ownedFIPConfigs, toDeleteConfigs, changedTotal, err
+			}
+			newConfigs[fipIsIPv6] = append(newConfigs[fipIsIPv6], fip)
+		}
+	}
+
 	handleFrontendIPConfig := func(isIPv6 bool) error {
-		ownedFIPConfig, toDeleteConfigsSingleStack, changed, newConfigs, err := az.reconcileFrontendIPConfigsSingleStack(clusterName, service, lb, lbStatus, wantLb, isIPv6, lbFrontendIPConfigName[isIPv6], &subnet)
+		ownedFIPConfig, toDeleteConfigsSingleStack, changed, newConfigs, err := az.reconcileFrontendIPConfigsSingleStack(clusterName, service, lb, lbStatus, wantLb, isIPv6, newConfigs[isIPv6], lbFrontendIPConfigName[isIPv6], &subnet)
```
Comment (nilo19) [inline pkg/provider/azure_loadbalancer.go:1840]: Since we have had lb here in the parameters, why bother to build a list fip configs outside of the function?

File: pkg/provider/azure_loadbalancer.go (hunk 3 above)
Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go:1861]: As I commented above, else means "we still need to put FIPs of that IP family into lbFrontEndIPConfigNewConfigs".

File: pkg/provider/azure_loadbalancer.go (hunk 4 above)
Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go:1868]: How about rename the var to something like "newConfigsWithIPFamily"? There're a lot of maps with isIPv6 as key and do you suggest changing them all?

File: pkg/provider/azure_loadbalancer.go (hunk 5 above)
Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go:1840]: Because `vXEnabled` -> `handleFrontendIPConfig()` -> `reconcileFrontendIPConfigsSingleStack()`.
If the Service is single stack, we need to handle the `newConfigs` of the other IP family outside.

File: pkg/provider/azure_utils.go (hunk 2 above)
Comment (lzhecheng) [reply pkg/provider/azure_utils.go]: Updated.

File: pkg/provider/azure_utils.go (hunk 1 above)
Comment (lzhecheng) [reply pkg/provider/azure_utils.go:451]: Updated.

File: pkg/provider/azure_loadbalancer.go (hunk 4 above)
Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go:1868]: Discussed offline. Will create a PR and set a const var for IsIPv6 and IsIPv4.

File: pkg/provider/azure_utils.go [hunk 6]
```diff
@@ -447,36 +447,28 @@ func getResourceByIPFamily(resource string, isDualStack, isIPv6 bool) string {
...
-		if !strings.EqualFold(fipPIPID, id) {
-			continue
+		fipPIPID := pointer.StringDeref(fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress.ID, "")
+		fipPIPName := fipPIPID[strings.LastIndex(fipPIPID, "/")+1:]
+		pip, exist, err := az.getPublicIPAddress(rg, fipPIPName, azcache.CacheReadTypeDefault)
+		if err != nil {
+			return false, err
 		}
-		if pip.PublicIPAddressPropertiesFormat != nil {
+		if exist && pip.PublicIPAddressPropertiesFormat != nil {
```
Comment (nilo19) [inline pkg/provider/azure_utils.go:471]: Is it possible that the ipv6 public IP has not yet been created?

File: pkg/provider/azure_utils.go [hunk 7]
```diff
@@ -485,11 +477,9 @@ func (az *Cloud) isFIPIPv6(service *v1.Service, fip *network.FrontendIPConfigura
 				return net.ParseIP(pointer.StringDeref(pip.PublicIPAddressPropertiesFormat.IPAddress, "")).To4() == nil, nil
 			}
 		}
-		klog.Errorf("Checking IP Family of PIP %q of corresponding frontend IP configuration %q of external Service but "+
-			"it is not clear. It's considered to be IPv4",
-			pointer.StringDeref(pip.Name, ""), pointer.StringDeref(fip.Name, ""))
-		break
 	}
+
+	klog.Errorf("Checking IP Family of frontend IP configuration %q but it is not clear. It's considered to be IPv4", pointer.StringDeref(fip.Name, ""))
```
Comment (nilo19) [inline pkg/provider/azure_utils.go:482]: It should be Warningf here.

Comment (lzhecheng) [reply pkg/provider/azure_utils.go]: Updated.

**Comment by lzhecheng** [comment]
/hold
Will manually check ccm log after all tests finish to see if isFIPIPv6 is working properly.

File: pkg/provider/azure_utils.go (hunk 6 above)
Comment (lzhecheng) [reply pkg/provider/azure_utils.go:476]: It won't. One evidence is that the warning in `isFIPIPv6()` shouldn't appear in CCM log: "Checking IP Family of frontend IP configuration %q but it is not clear. It's considered to be IPv4".

File: pkg/provider/azure_loadbalancer_test.go [hunk 8]
```diff
@@ -6730,6 +6708,112 @@ func TestReconcileZonesForFrontendIPConfigs(t *testing.T) {
...
+	testcases := []struct {
+		desc          string
+		service       v1.Service
+		existingFIPs  []network.FrontendIPConfiguration
+		existingPIPs  []network.PublicIPAddress
+		status        *v1.LoadBalancerStatus
+		expectedDirty bool
+		expectedFIPs  []network.FrontendIPConfiguration
+	}{
+		{
```
Comment (nilo19) [inline pkg/provider/azure_loadbalancer_test.go:6724]: can you add a test case for reconciling existing configs owned by the service without touching the others?

Comment (nilo19) [reply pkg/provider/azure_loadbalancer_test.go:6724]: also add some negative cases please

File: pkg/provider/azure_utils.go [hunk 9]
```diff
@@ -448,50 +448,81 @@ func getResourceByIPFamily(resource string, isDualStack, isIPv6 bool) string {
...
+		return service.Spec.IPFamilies[0] == v1.IPv6Protocol, nil
 	}
-	var fipPIPID string
-	if fip.FrontendIPConfigurationPropertiesFormat != nil && fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress != nil {
-		fipPIPID = pointer.StringDeref(fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress.ID, "")
+
+	// First check private
+	// For an external LB, the FIP's private IP address is empty. The logic will
+	// go through this part and check public IP address then.
+	if fip.FrontendIPConfigurationPropertiesFormat != nil {
```
Comment (nilo19) [inline pkg/provider/azure_utils.go:470]: I checked all the refs of this function. If we can make sure the fip is always owned by the service, the pip check is not needed.

File: pkg/provider/azure_loadbalancer_test.go (hunk 8 above)
Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer_test.go:6726]: UT added.

File: pkg/provider/azure_utils.go (hunk 9 above)
Comment (lzhecheng) [reply pkg/provider/azure_utils.go:470]: The FIP also needs to be `isPrimaryService==true`. If it is the secondary Service, I think the pip check is still needed.

Comment (nilo19) [reply pkg/provider/azure_utils.go:470]: If the fip is owned by a single stack service, we can check the service instead of pip, no matter if it is primary or secondary.

Comment (lzhecheng) [reply pkg/provider/azure_utils.go:475]: It makes sense for single stack svc. Updated.

**Comment by feiskyer** [comment]
> Will manually check ccm log after all tests finish to see if isFIPIPv6 is working properly.

Have the test finished?

**Comment by lzhecheng** [comment]
/test pull-cloud-provider-azure-e2e-ccm-ipv6-capz
/test pull-cloud-provider-azure-e2e-ccm-ipv6-vmss-capz

@feiskyer the tests are manually triggered like above. Those runs before were successful

File: pkg/provider/azure_loadbalancer.go [hunk 10]
```diff
@@ -1239,30 +1236,32 @@ func (az *Cloud) isFrontendIPChanged(
...
 	loadBalancerIP := getServiceLoadBalancerIP(service, isIPv6)
+	isInternal := requiresInternalLoadBalancer(service)
 	if isInternal {
 		// Judge subnet
 		subnetName := getInternalSubnet(service)
 		if subnetName != nil {
-			if err := az.fillSubnet(subnet, *subnetName); err != nil {
-				return false, err
+			if subnet == nil {
+				return false, fmt.Errorf("isFrontendIPChanged: Unexpected nil subnet")
```
Comment (jwtty) [inline pkg/provider/azure_loadbalancer.go:1263]: nit: can we include subnet name in the error msg?

Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go:1263]: Let's do it in the next PR.
